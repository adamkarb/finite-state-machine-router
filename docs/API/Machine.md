# `<Machine/>`
`<Machine/>` component is the top-lever wrapper compnoent. It contains unique naming and routing information for the rest of the state machine. The `<Machine/>` component can only ever have one child active at a time.

## Properties
&ast; = Required property
<table>
    <tr>
        <th>Property</th>
        <th>Type</th>
        <th>Default</th>
        <th>Description</th>
    </tr>
    <tr>
        <td><code>history</code>*</td>
        <td><code>Object</code></td>
        <td></td>
        <td><code>History</code> package instance that will handle history navigation. This can be either <code>createBrowserHistory</code> or <code>createMemoryHistory</code>. Check out the [History package documentation](ttps://github.com/ReactTraining/history/blob/master/docs/api-reference.md).</td>
    </tr>
    <tr>
        <td><code>id</code>*</td>
        <td><code>String</code></td>
        <td></td>
        <td>Unique identifier for this machine. This must be different than all ancestors and siblings.</td>
    </tr>
    <tr>
        <td><code>logger</code>*</td>
        <td><code>Boolean</code></td>
        <td><code>false</code></td>
        <td>Enables console logging for easier debugging. Enabled for <code>transitions</code> and <code>history</code> changes.</td>
    </tr>
    <tr>
        <td><code>path</code></td>
        <td><code>String</code></td>
        <td></td>
        <td>The URL that the browser will resolve to upon entry of this machine. It is relative to all ancestor urls.</td>
    </tr>
</table>

## Reference
[SCXML W3C Specification - SCXML](https://www.w3.org/TR/scxml/#scxml)
>The top-level wrapper element, which carries version information. The actual state machine consists of its children. Note that only one of the children is active at any one time.