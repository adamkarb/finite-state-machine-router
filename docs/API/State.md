# `<State/>`
`<State/>` components are the building block of our application. Everything our user will see is a combination of states, with most supplied with React Components to render. `<State/>` components contain unique naming and routing information for the entire state machine to follow when applicable.

## Properties
&ast; = Required property
<table>
    <tr>
        <th>Property</th>
        <th>Type</th>
        <th>Default</th>
        <th>Description</th>
    </tr>
    <tr>
        <td><code>component</code></td>
        <td><code>React component</code></td>
        <td></td>
        <td>React component to be rendered in this state.</td>
    </tr>
    <tr>
        <td><code>id</code>*</td>
        <td><code>String</code></td>
        <td></td>
        <td>Unique identifier for this state. This must be different than all ancestors and siblings.</td>
    </tr>
    <tr>
        <td><code>initial</code></td>
        <td><code>Boolean</code></td>
        <td><code>false</code></td>
        <td>Denotes the state to be resolved to upon state machine initiation. If no children states are marked as initial, the first child state will be rendered. Doesn't apply if the parent state is a parallel type, because all children of parallel type are simultaneously active.</td>
    </tr>
    <tr>
        <td><code>path</code></td>
        <td><code>String</code></td>
        <td></td>
        <td>The URL that the browser will resolve to upon entry of this state. It is relative to all ancestor urls.</td>
    </tr>
</table>

## Reference
[SCXML W3C Specification - State](https://www.w3.org/TR/scxml/#state)
>Holds the representation of a state.