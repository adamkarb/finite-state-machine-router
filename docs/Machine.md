# Machine
**Reference:** [SCXML W3C Specification - SCXML](https://www.w3.org/TR/scxml/#state)

"The top-level wrapper element, which carries version information. The actual state machine consists of its children. Note that only one of the children is active at any one time."

## State resolution
Upon initialization of our `Machine` wrapper component, the active child state is determined in this order:
- If the URL pathname (such as `//localhost:3000/example/pathname)` is present (obtained from (History)[https://github.com/ReactTraining/history/blob/master/docs/GettingStarted.md#properties] package's `history.location.pathname`), the children states with matching `url` properties will be resolve until the URL pathname tree is depleted. If the resultant state is not of type `atomic`, then each `initial` child state is resolved.
- If a URL pathname is not present or does not apply, each `initial` child state is resolved until a state of type `atomic` is reached.

## Signature
`<Machine {...props} />`

## Properties
&ast; = Required property
<table>
    <tr>
        <th>Property</th>
        <th>Type</th>
        <th>Default</th>
        <th>Description</th>
    </tr>
    <tr>
        <td><code>id</code>*</td>
        <td><code>String</code></td>
        <td></td>
        <td>Unique identifier for this machine. This must be different than all ancestors and siblings.</td>
    </tr>
    <tr>
        <td><code>url</code></td>
        <td><code>String</code></td>
        <td></td>
        <td>The URL that the browser will resolve to upon entry of this machine. It is relative to all ancestor urls.</td>
    </tr>
</table>

## Children
- [<State ... />](./State.md)