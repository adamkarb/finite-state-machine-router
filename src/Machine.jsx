import React, { useContext, useEffect, useMemo, useState } from 'react';
import { createBrowserHistory } from 'history';
import {
    getChildStateNodes,
    // getInitialChildStateNode,
    injectUrlParams,
    normalizeChildStateProps,
    resolveUrlToAtomic,
    getAtomic,
    selectTransition
} from './util';
import { logger } from './logger';

export const MachineContext = React.createContext({});
MachineContext.displayName = 'Machine';

export const createMachine = (options) => (props) => Machine({ ...props, ...options });

export const useMachine = () => {
    const { current, history, id, params, send } = useContext(MachineContext);
    return [{ current, history, id, params }, send ];
}

function Machine ({ children: machineChildren, history: machineHistory, id: machineId = 'machine', logging = false, path: machinePath }) {
    const history = useMemo(() => machineHistory || createBrowserHistory({ basename: machinePath }), []);

    const [ childStates, normalized ] = useMemo(() => {
        const _childStates = getChildStateNodes(React.Children.toArray(machineChildren));

        if (_childStates.length === 0) {
            throw new Error('<Machine/> has no children <State/> nodes! At least one is required to be considered a valid state machine.');
        }

        const _normalized = normalizeChildStateProps(_childStates, machineId);

        return [ _childStates, _normalized ];
    }, [ machineChildren ]);

    const [ initialStack, params, path ] = useMemo(() => {
        const { params, path, stack, url } = resolveUrlToAtomic(history.location.pathname, normalized, machineId);

        if (history.location.pathname !== url) {
            history.replace(url);
        }

        return [ stack, params, path ];
    }, []);

    const [ state, setState ] = useState({
        current: initialStack,
        params,
        path
    });

    const resolvePath = (path, params, source, target) => {
        const url = injectUrlParams(path, params);

        if (url !== history.location.pathname) {
            history.push(url, {
                source,
                target
            });
        }
    }

    const send = (event, data = null) => {
        const targetState = selectTransition(event, state.current, normalized);

        if (targetState) {
            const params = data && data.params || state.params;
            const { cond, event: transitionEvent, target: targetId } = targetState;
            const targetNode = normalized.find(norm => norm.id === targetId);

            if (targetNode) {
                const { path, stack } = getAtomic(targetNode.stack, normalized);

                logging && logger({
                    action: 'TRANSITION',
                    data,
                    event,
                    source: {
                        state: state.current,
                        path: state.path
                    },
                    target: {
                        state: stack,
                        path
                    }
                });

                setState({ current: stack, params, path });
                resolvePath(path, params, state.current, stack);
            } else {
                // console.error(`Invalid transition target: No target State Node of id "${targetId}" exists. event ${event} will be discarded.`);
                logging && logger({
                    action: 'EVENT_DISCARDED',
                    data,
                    event,
                    reason: 'NO_MATCHING_STATE',
                    source: {
                        state: state.current,
                        path: state.path
                    },
                    target: {
                        state: targetId   
                    }
                });
            }
        } else {
            logging && logger({
                action: 'EVENT_DISCARDED',
                data,
                event,
                reason: 'NO_MATCHING_TRANSITION',
                source: {
                    state: state.current,
                    path: state.path
                },
            });
        }
    }

    useEffect(() => history.listen(({ action, location }) => {
        if ((!location.state || !location.state.target) || action === 'POP') {
            const { params, path, stack, url } = resolveUrlToAtomic(location.pathname, normalized, machineId);

            logging && logger({
                action: 'HISTORY_CHANGE',
                data: location.pathname,
                source: {
                    state: state.current,
                    path: state.path
                },
                target: {
                    state: stack,
                    path
                }
            });

            setState({ current: stack, params, path });
        }
    }), [ history.location.pathname ]);

    const providerValue = {
        ...state,
        history,
        id: machineId,
        send
    };

    return <MachineContext.Provider value={providerValue}>
        {childStates}
    </MachineContext.Provider>;
}

Machine.displayName = 'Machine';

// export const createMachine = (props) => Machine(props);
export default Machine;
