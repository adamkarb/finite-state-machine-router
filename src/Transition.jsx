const Transition = (props) => null;
Transition.displayName = 'Transition';

export default Transition;
