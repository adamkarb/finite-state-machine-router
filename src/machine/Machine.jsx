import React, { useEffect, useMemo, useState } from 'react';
import { createBrowserHistory } from 'history';
import {
    deriveStateFromUrl,
    getChildStateNodes,
    getInitialChildStateNode,
    injectUrlParameters,
    normalizeChildStateProps,
    resolveInitialStack,
    selectTransition
} from './util';

export const MachineContext = React.createContext();
MachineContext.displayName = 'Machine';

export function Machine ({ children: machineChildren, history, id: machineId, path: machinePath }) {
    // Default history
    if (!history) {
        history = createBrowserHistory({ basename: machinePath });
    }

    // Resolve from URL -> input URL -> get stack -> resolve to initial atomic
    // Resolve from state -> input state ID -> get stack -> resolve to initial atomic

    const { childStates, initialStack, normalized, params } = useMemo(() => {
        let childStates = getChildStateNodes(machineChildren);
        let { pathname } = history.location;
        let initialStack = '#' + machineId + '.' + getInitialChildStateNode(childStates).props.id;
        const normalized = normalizeChildStateProps(childStates, machineId);
        const { params, path: currentPath, stack: currentStack } = deriveStateFromUrl(pathname, normalized, machineId);
        const { route, stack } = resolveInitialStack(currentStack, normalized);

        const initialRoute = injectUrlParameters(route, params);
        history.push(initialRoute);

        return {
            childStates,
            initialStack: stack || initialStack,
            normalized,
            params
        }
    }, [ machineChildren ]);

    const [ state, setState ] = useState(initialStack);

    function resolvePath(path) {
        const url = injectUrlParameters(path, params);

        if (url !== history.location.pathname) {
            console.log('resolvePath', history.location.pathname, 'to', url);
            history.push(url, { stack: state });
        }
    }

    function resolveState(stateId) {
        const { route, stack } = resolveInitialStack(
            normalized.find(norm => norm.id === stateId).stack,
            normalized
        );
        console.log('resolveState', state, '->', stack);
        setState(stack);
    }

    function send(event, data = null) {
        console.log('send', event, data);
        const { cond, event: transitionEvent, target } = selectTransition(event, state, normalized);
        console.log(target);
        resolveState(target);
        // setEvent({ event, ...data });
    }

    useEffect(() => history.listen((location, action) => {
        const { pathname } = location;
        const { params, path: currentPath, stack: currentStack } = deriveStateFromUrl(pathname, normalized, machineId);
        const { route, stack } = resolveInitialStack(currentStack, normalized);

        if (stack !== state) {
            console.log('yep', stack, state);
            setState(stack);
        } else {
            resolvePath(route);
        }
    }));

    // useEffect(() => {
    //     console.log('useEffect', state);
    // });
    
    // console.log('render', machineId);

    const providerValue = {
        current: state,
        // event,
        history,
        id: machineId,
        params,
        resolvePath,
        // resolveState,
        send
    }

    return <MachineContext.Provider value={providerValue}>
        {childStates}
    </MachineContext.Provider>;
}
