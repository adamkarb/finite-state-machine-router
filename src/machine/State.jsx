import React, { useContext, useEffect, useMemo, useState } from 'react';
import { MachineContext } from './Machine';
import { getChildrenOfType, getChildStateNodes, isCurrentStack, isExactStack } from './util';

export const StateNodeContext = React.createContext({
    id: null,
    path: null,
    stack: null
});
StateNodeContext.displayName = 'StateNode';

function State(props) {
    const { children, component: Component, id, initial, invoke, onEntry, onExit, path, type } = props;
    const machineContext = useContext(MachineContext);
    const { event: machineEvent, current, history, id: machineId, params, resolvePath, send: machineSend } = machineContext;
    const { id: parentId, path: parentPath, stack: parentStack } = useContext(StateNodeContext);
    const stack = parentStack ? `${parentStack}.${id}` : `#${machineId}.${id}`;
    const stackPath = path ? parentPath ? parentPath + path : path : parentPath;
    const match = isCurrentStack(id, current) ? {
        exact: isExactStack(id, current),
        params,
        path: stackPath,
        url: history.location.pathname
    } : false;

    const [ _type, transitions ] = useMemo(() => {
        let _type = type;
        const childStates = getChildStateNodes(children);
        const transitions = getChildrenOfType(children, 'Transition').map(({ props }) => ({
            cond: props.cond || null,
            event: props.event,
            sendid: id,
            target: props.target,
            type: 'internal'
        }));
        // const initialChild = childStates.find(c => c.props.initial) || childStates[0];

        if (_type !== 'parallel') {
            if (!childStates.length) {
                _type = 'atomic';
            } else if (childStates.length > 1) {
                _type = 'compound';
            } else {
                _type = 'default';
            }
        }

        return [ _type, transitions ];
    }, [ children ]);

    useEffect(() => {
        if (match) {
            if (_type === 'atomic') {
                // console.log('resolvepath');
                stackPath ? resolvePath(stackPath) : resolvePath('/');
            }
        }
    }, [ current ]);

    useMemo(() => {
        if (match && invoke) {
            machineSend('invoke.' + id);
            invoke && invoke(machineContext);
        }
    }, []);

    const initialContext = {
        id,
        path: stackPath,
        stack,
        send: machineSend,
    }
    const componentProps = {
        children,
        history,
        machine: {
            current,
            // transitions,
            send: machineSend
        },
        match
    }

    return match ?
        <StateNodeContext.Provider value={initialContext}>
            { Component ?
                <Component {...componentProps}/>
            : children }
        </StateNodeContext.Provider>
    : null;
}

export default State;
