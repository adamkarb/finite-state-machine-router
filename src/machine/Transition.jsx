import { useContext, useEffect } from 'react';
import { MachineContext } from './Machine';

function Transition({ event, target }) {
    const machineContext = useContext(MachineContext);
    const { _event: machineEvent, resolveState } = machineContext;

    // get state node stack, send as target in "machienSend"
    // what are the limitations on "target" prop? sibling? ancestor? descendent?

    useEffect(() => {
        // console.log('transition');

        if (machineEvent && machineEvent.name === event) {
            console.log('transition', machineEvent, target);
            resolveState(target);
        }
    }, [ machineEvent ]);

    return null;
}

export default Transition;