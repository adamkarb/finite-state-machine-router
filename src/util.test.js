import React from 'react';
import Machine from './Machine';
import State from './State';
import Transition from './Transition';
import {
    // getChildrenOfType,
    classNames,
    getChildStateNodes,
    getInitialChildStateNode,
    injectUrlParams,
    isCurrentStack,
    isExactStack,
    normalizeChildStateProps,
    resolveInitial,
    resolveToAtomic,
    selectTransition
} from './util';

describe('utility functions', () => {
    const machine = <Machine id='home' path='/home'>
            <State id='parent' path='/:parent'>
            <State id='child-1' path='/child-1'>
                <Transition event='test-event' target='child-2'/>
                <State id='grand-child'/>
            </State>
            <State id='child-2'/>
        </State>
    </Machine>;
    const normalized = normalizeChildStateProps(React.Children.toArray(machine.props.children), 'home');

    test('classNames', () => { 
        const className = [
            'custom-class',
            { 'should-display': true },
            { 'should-not-display': false }
        ];

        const className2 = [
            { 'should-not-display': false }
        ];

        expect (classNames(className)).toBe('custom-class should-display');
        expect (classNames(className2)).toBeNull();
    });

    test('getChildStateNodes', () => {
        const childStates = getChildStateNodes(machine.props.children).every(child => child.type.displayName === 'State');
        expect(childStates).toBe(true);
    });

    test('getInitialChildStateNode', () => {
        const machineWithoutInitial = [
            <State id='child-1'/>,
            <State id='child-2'/>
        ];

        const machineWithInitial = [
            <State id='child-1'/>,
            <State id='child-2' initial/>
        ];

        const result1 = getInitialChildStateNode(getChildStateNodes(machineWithoutInitial)).props.id;
        const result2 = getInitialChildStateNode(getChildStateNodes(machineWithInitial)).props.id;

        expect(result1).toBe('child-1');
        expect(result2).toBe('child-2');
    });

    test('injectUrlParams', () => {
        const params = {
            'dynamicPath': 'dynamic-path',
            'anotherDynamicPath': 'another-dynamic-path'
        };

        expect(injectUrlParams('/:dynamicPath', params)).toBe('/dynamic-path');
        expect(injectUrlParams('/:dynamicPath/:anotherDynamicPath', params)).toBe('/dynamic-path/another-dynamic-path');
        expect(injectUrlParams('/:dynamicPath/static-path/:anotherDynamicPath', params)).toBe('/dynamic-path/static-path/another-dynamic-path');
        expect(injectUrlParams('/:dynamicPath/static-path', params)).toBe('/dynamic-path/static-path');
        expect(injectUrlParams('/static-path/:dynamicPath', params)).toBe('/static-path/dynamic-path');
        expect(injectUrlParams('/static-path/:dynamicPath/another-static-path', params)).toBe('/static-path/dynamic-path/another-static-path');
    });

    test('isCurrentStack', () => {
        expect(isCurrentStack('child-1', '#home.parent.child-1.grand-child')).toBe(true);
        expect(isCurrentStack('child-2', '#home.parent.child-1.grand-child')).toBe(false);
        expect(isCurrentStack('grand-child', '#home.parent.child-1.grand-child')).toBe(true);
    });

    test('isExactStack', () => {
        expect(isExactStack('child-1', '#home.parent.child-1.grand-child')).toBe(false);
        expect(isExactStack('child-2', '#home.parent.child-1.grand-child')).toBe(false);
        expect(isExactStack('grand-child', '#home.parent.child-1.grand-child')).toBe(true);
    });

    test('normalizeChildStateProps', () => expect(normalized).toMatchSnapshot());

    test('resolveInitial', () => {
        expect(resolveInitial('/parent-id/child-1', normalized, 'home')).toMatchSnapshot();
    })

    test('resolveToAtomic', () => {
        expect(resolveToAtomic('#home.parent', normalized)).toMatchSnapshot();
        expect(resolveToAtomic('#home.parent.child-1', normalized)).toMatchSnapshot();
        expect(resolveToAtomic('#home.parent.child-2', normalized)).toMatchSnapshot();
    });

    test('selectTransition', () => {
        expect(selectTransition('no-matching-event', '#home.parent.child-1', normalized)).toBe(null);
        expect(selectTransition('test-event', '#home.parent.child-1', normalized)).toMatchSnapshot();
    });
});
