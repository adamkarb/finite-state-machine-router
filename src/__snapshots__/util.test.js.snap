// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utility functions normalizeChildStateProps 1`] = `
Array [
  Object {
    "childStates": Array [
      "child-1",
      "child-2",
    ],
    "id": "parent",
    "initial": true,
    "path": "/:parent",
    "stack": "#home.parent",
    "transitions": Array [],
    "type": "compound",
  },
  Object {
    "childStates": Array [
      "grand-child",
    ],
    "id": "child-1",
    "initial": true,
    "path": "/:parent/child-1",
    "stack": "#home.parent.child-1",
    "transitions": Array [
      Object {
        "cond": null,
        "event": "test-event",
        "sendId": "child-1",
        "target": "child-2",
      },
    ],
    "type": "default",
  },
  Object {
    "childStates": Array [],
    "id": "grand-child",
    "initial": true,
    "path": "/:parent/child-1",
    "stack": "#home.parent.child-1.grand-child",
    "transitions": Array [],
    "type": "atomic",
  },
  Object {
    "childStates": Array [],
    "id": "child-2",
    "initial": false,
    "path": "/:parent",
    "stack": "#home.parent.child-2",
    "transitions": Array [],
    "type": "atomic",
  },
]
`;

exports[`utility functions resolveInitial 1`] = `
Object {
  "params": Object {
    "parent": "parent-id",
  },
  "path": "/:parent/child-1",
  "stack": "#home.parent.child-1.grand-child",
  "url": "/parent-id/child-1",
}
`;

exports[`utility functions resolveToAtomic 1`] = `
Object {
  "path": "/:parent/child-1",
  "stack": "#home.parent.child-1.grand-child",
}
`;

exports[`utility functions resolveToAtomic 2`] = `
Object {
  "path": "/:parent/child-1",
  "stack": "#home.parent.child-1.grand-child",
}
`;

exports[`utility functions resolveToAtomic 3`] = `
Object {
  "path": "/:parent",
  "stack": "#home.parent.child-2",
}
`;

exports[`utility functions selectTransition 1`] = `
Object {
  "cond": null,
  "event": "test-event",
  "sendId": "child-1",
  "target": "child-2",
}
`;
